include/private/gc_priv.h

```diff
@@ -2116,11 +2116,14 @@ GC_INNER void GC_with_callee_saves_pushed(void (*fn)(ptr_t, void *),
   /* Load value, check tag and permissions of the target memory.   */
 # define LOAD_WORD_OR_CONTINUE(v, p) \
         { \
-            v = *(void **)(p);                                            \
-            size_t has_rwx = cheri_perms_get(v) & (CHERI_PERM_LOAD        \
-                                                   | CHERI_PERM_STORE     \
-                                                   | CHERI_PERM_EXECUTE); \
-            if ((cheri_tag_get(v) == 0) || !has_rwx) continue;            \
+            v = *(void **)(p);                                               \
+            size_t in_bounds = (v >= (word)cheri_base_get(v))                \
+                                && (v < (word)cheri_base_get(v)              \
+                                        + (word)cheri_length_get(v));        \
+            size_t has_rwx = cheri_perms_get(v) & (CHERI_PERM_LOAD           \
+                                                   | CHERI_PERM_STORE        \
+                                                   | CHERI_PERM_EXECUTE);    \
+            if ((cheri_tag_get(v) == 0) || !in_bounds || !has_rwx) continue; \
```

Comment:
I think the `cheri_tag_get` check should go at the top of the function. [Also `if (!cheri_tag_get(v))` is probably easier than `if ((cheri_tag_get(v) == 0)`.

Reply:
Yes, IMHO, two `continue`s are better here, because now an attacker (with an invalid capability) can't trick you into undefined behaviour -- that's a win in my book :)

Reply:
Not sure I understand.  
In the original code, the tag is checked first and if true, the `if` condition is short-circuited. Permission and bounds are not even required for `continue` to be triggered.  If there is an invalid capability with correctly crafted bounds and permissions,  the tag is still getting checked within the original condition and the continue will be triggered. 

Is there some corner condition that I am not seeing ?

Reply:
The tag check comes after:

```
            size_t in_bounds = (v >= (word)cheri_base_get(v))                \
                                && (v < (word)cheri_base_get(v)              \
                                        + (word)cheri_length_get(v));        \
```

I believe that code could cause undefined behaviour with an invalid capability. The tag check must therefore come before it.

Reply:
Wasn't the undefined behaviour for **set()**ting the bounds and not **get()**ting it. I checked that you can  get the bounds on an invalid capability. I checked on both *riscv64-purecap* and *morello-purecap* .  Just asking purely with a "curiosity killed the cat" intent.  Otherwise no problem with making the change

Reply:
If you do `a + b` and you are not in control of one (or both) of `a` and `b`, then the addition can cause undefined behaviour.

Reply:
But isn't that what the `cheri_tag_get()` at the start of the condition  is supposed to prevent ? if `v` is an invalid capability, regardless of any junk that might result from extracting and checking  the permissions or calculating the bounds,  the `!cheri_tag_get()` will  be true and cause the capability to not be dereferenced ?

The change is made in [07b586b](https://github.com/capablevms/bdwgc/pull/19/commits/07b586b7ec011b520062977bd8e863a51061e711)